onnxruntime/core/providers/nnapi/nnapi_builtin/builders/impl/reduction_op_builder.cc

```diff
@@ -0,0 +1,234 @@
+  const auto& output = node_unit.Outputs()[0].node_arg.Name();
+
+  auto& shaper(model_builder.GetShaper());
+  const auto input_shape = shaper[inputs[0].node_arg.Name()];
+  const auto& operand_indices(model_builder.GetOperandIndices());
+  const auto& operand_types(model_builder.GetOperandTypes());
+
+  NodeAttrHelper helper(node_unit);
+
+  int32_t op_code;
+  if (op_type == "ReduceMean") {
+    op_code = ANEURALNETWORKS_MEAN;
+  } else {
+    // TODO: Add more reduction ops support
+    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "ReductionOpBuilder, unknown op: ", op_type);
+  }
+
+  if (op_type == "ReduceMean") {
+    // Get axes for ReduceMean
+    std::vector<int32_t> axes;
+    if (inputs.size() > 1 && inputs[1].node_arg.Exists()) {
+      // ReduceMean-18 uses the second optional input as axes
+      const auto& initializers(model_builder.GetInitializerTensors());
+      const auto& axes_tensor = *initializers.at(inputs[1].node_arg.Name());
+      Initializer unpacked_tensor(axes_tensor);
+      auto raw_axes = unpacked_tensor.DataAsSpan<int64_t>();
+      const auto size = SafeInt<uint32_t>(axes_tensor.dims()[0]);
+      axes.resize(size);
+      for (uint32_t i = 0; i < size; i++) {
+        // it is unlikely we have an axis value overflow for int32
+        axes[i] = static_cast<int32_t>(raw_axes[i]);
+      }
+    } else if (helper.HasAttr("axes")) {
+      // For ReduceMean-13 or eariler, retrieve axes from the attribute
+      const auto& axes_int64 = helper.Get("axes", std::vector<int64_t>{});
+      axes.reserve(axes_int64.size());
+      for (auto& axis : axes_int64) {
+        axes.push_back(static_cast<int32_t>(axis));
+      }
+    }
+
+    // Add ReduceMean operation
+    const auto keepdims = helper.Get("keepdims", 1);
+
+    InlinedVector<uint32_t> input_indices;
+    input_indices.push_back(operand_indices.at(inputs[0].node_arg.Name()));  // data
+
+    if (!axes.empty()) {
+      for (auto& axis : axes) {
+        axis = static_cast<int32_t>(HandleNegativeAxis(axis, input_shape.size()));
+      }
+
+      const auto axes_name = model_builder.GetUniqueName(node_unit.Name() + inputs[0].node_arg.Name() + "_axes");
+      Shape axes_dimen = {static_cast<uint32_t>(axes.size())};
+      const OperandType axes_operand_type(Type::TENSOR_INT32, axes_dimen);
+      ORT_RETURN_IF_ERROR(model_builder.AddOperandFromPersistMemoryBuffer(axes_name, axes.data(), axes_operand_type));
+
+      input_indices.push_back(operand_indices.at(axes_name));  // axes
+
+      int32_t input_size = static_cast<int32_t>(input_shape.size());
+      std::unordered_set<int32_t> axes_to_be_reduced;
+
+      for (const auto& axis : axes) {
+        axes_to_be_reduced.insert(axis);
+      }
+
+      // Make output dimensions
+      InlinedVector<uint32_t> output_dimen;
+      if (keepdims == 1) {
+        output_dimen.reserve(input_size);
+      } else {
+        output_dimen.reserve(input_size - axes_to_be_reduced.size());
```

Comment:
Do we need  to add a `squeeze` OP here?

Reply:
sorry - can you elaborate more about why we need a squeeze here?

I think only when the NNAPI shape output behavior does not match with the ONNX one, then we need to make changes. for instance,

in Gemm op case, we would want to squeeze the `bias` input as onnx allow >1D bias tensor but NNAPI accepts 1D bias input, so we would want to squeeze bias is needed. https://github.com/microsoft/onnxruntime/blob/0d6416c0e9eb9ba537890d0fe3f695693a08a8fe/onnxruntime/core/providers/nnapi/nnapi_builtin/builders/impl/gemm_op_builder.cc#L199-L204

And in your gather op case, I guess we changed the behavior now to have an extra {1} in dim, so we need to squeeze it back. 

Correct me if I am wrong.

Reply:
I see.  Reduce_xxx in NNAPI does support attribute "keepdims".